src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs

```diff
@@ -0,0 +1,473 @@
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(17)]
+        [InlineData(uint.MaxValue)]
+        public static void TestUInt32(uint value)
+        {
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetUInt32(value);
+            Assert.Equal(value, jsonNumber.GetUInt32());
+            Assert.True(jsonNumber.TryGetUInt32(out uint result));
+            Assert.Equal(value, result);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(17)]
+        [InlineData(ulong.MaxValue)]
+        public static void TestUInt64(ulong value)
+        {
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetUInt64(value);
+            Assert.Equal(value, jsonNumber.GetUInt64());
+            Assert.True(jsonNumber.TryGetUInt64(out ulong result));
+            Assert.Equal(value, result);
+        }
+
+        [Fact]
+        public static void TestDecimal()
+        {
+            var decimalValue = decimal.MaxValue;
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetDecimal(decimalValue);
+            Assert.Equal(decimalValue, jsonNumber.GetDecimal());
+            Assert.True(jsonNumber.TryGetDecimal(out decimal result));
+            Assert.Equal(decimalValue, result);
+        }
+
+        [Theory]
+        [InlineData("3.14")]
+        [InlineData("-17")]
+        [InlineData("0")]
+        [InlineData("189")]
+        [InlineData("1e400")]
+        [InlineData("1e+100000002")]
+        [InlineData("184467440737095516150.184467440737095516150")]
+        public static void TestString(string value)
+        {
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetFormattedValue(value);
+            Assert.Equal(value, jsonNumber.ToString());
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("3,14")]
+        [InlineData("this is not a number")]
+        [InlineData("NAN")]
+        [InlineData("0.")]
+        [InlineData("008")]
+        [InlineData("0e")]
+        [InlineData("5e")]
+        [InlineData("5a")]
+        [InlineData("0.1e")]
+        [InlineData("-01")]
+        [InlineData("10.5e")]
+        [InlineData("10.5e-")]
+        [InlineData("10.5e+")]
+        [InlineData("10.5e-0.2")]
+        [InlineData(" 6")]
+        [InlineData("6 ")]
+        [InlineData(" 6 ")]
+        [InlineData("+0")]
+        [InlineData("+1")]
+        public static void TestInvalidString(string value)
+        {
+            Assert.Throws<ArgumentException>(() => new JsonNumber(value));
+        }
+
+        [Fact]
+        public static void TestNullString()
+        {
+            Assert.Throws<ArgumentNullException>(() => new JsonNumber(null));
+        }
+
+        [Fact]
+        public static void TestIntFromString()
+        {
+            var jsonNumber = new JsonNumber("145");
+            Assert.Equal(145, jsonNumber.GetInt32());
+        }
+
+        [Fact]
+        public static void TestUpcasts()
```

Comment:
Will up-casting any JsonNumber always succeed? What about double <-> decimal. What's the up-cast relationship there? Or ulong <-> double?